design/one-pager-function-go-templating.md

```diff
@@ -0,0 +1,185 @@
+# Go Templating Function
+ 
+* Owner: Ezgi Demirel (@ezgidemirel)
+* Reviewers: Crossplane Maintainers
+* Status: Accepted
+
+## Background
+
+With Crossplane 1.14, Composition Functions will be promoted to `v1beta1`, and 
+several composition functions will be introduced as an extension point to be 
+used in pipelines. One of these extensions is the [Go Templating Function], 
+which allows users to render Crossplane resources using Go templating 
+capabilities like conditionals and loops. In addition to that, users can use 
+values in the environment configs or other resource fields to create/configure
+Crossplane resources.
+
+## Getting Template Files
+
+With the initial implementation of the Go Templating Function, users will be 
+able to provide templates in two different ways:
+- An `inline` source can be used to provide template files as a string in the
+  `input.inline.template` field of the pipeline step.
+- A `fileSystem` source can be used to provide template files with folder path
+  in the `input.fileSystem.dirPath` field of the pipeline step.
+
+Both of these options have their own advantages and disadvantages. While 
+`inline` source is faster and easier to use, it is not suitable for large 
+template files and hard to maintain. On the other hand, `fileSystem` source
+requires the files to be present in the local file system, which might be 
+possible by mounting config maps with `DeploymentRuntimeConfig` or building 
+another container image on top of the existing one with the template files. 
+
+In the next iterations, it is planned to support fetching template files from a
+remote location like GitHub repositories or S3 buckets.
```

Comment:
Do we plan to package templates to OCI images?

In the abstracted helm approach( which is kind of the contextual predecessor to the Go templating function) I was packaging helm chart together with the main Configuration and used the same tag to create a relation between them and then uptest them together.

So we have two OCI repos there
* platform-ref-apigateway for the main Configuration
* platform-ref-apigateway-chart for the Templates/Chart

This approach enabled me to create reproducible builds of both the Configuration and the Templates(Chart) that are used within this Configuration.

The main concern with just free style pulling templates from arbitrary places is that Templates will have totally different lifecycle from the main Composition. This concern is probably valid for the cases when we are mixing in together with the standard patch & transform function(similar to https://github.com/upbound/platform-ref-apigateway/blob/main/package/composition.yaml) or some other one. 

 In any case, testing Configuration that is pulling the logic from the remote sources does impose the testing and eventual reliability challenges as we might pull in some code that might lead to undesired side effects.

Reply:
I agree with your concerns. We'll enable so many different GitOps combinations for compositions. Even with config map mounting + Argo can cause unexpected side effects.

Reply:
> Do we plan to package templates to OCI images?

I like this, I believe this is a nice interim step between all inline templates and read from git which could probably give us what we want with the minimal effort and without dealing with all sort of complexities. 

We can have an API like:

```
      input:
        apiVersion: gotemplating.fn.crossplane.io/v1beta1
        kind: GoTemplate
        source: image
        image:
          repo: turkenh/my-templates
          tag: v0.1.0
```

and read templates inside the image from a predefined path, e.g. `/templates`. Coupling/decoupling the lifecycle of templates with the configuration package could be left to the author depending on their use case.

@ezgidemirel please open an issue for this as well 🙏